### PR TITLE
Integrate immutable-js-oss fork into main repository

### PR DESCRIPTION
**Pull request by jdeniau**
This PR does integrate the https://github.com/immutable-js-oss/immutable-js fork into the main repository main branch.

I think the only thing I did not import here is [Use ranges in devDependencies](https://github.com/immutable-js-oss/immutable-js/pull/192). Apparently @leebyron does not use ranges, when @Methuselah96 and I prefer to (I do not know for other maintainters). I preferred keep the actual preferences here, but It may be discussed in another PR.

A (possibly sensible) think to discuss too I think is the [FUNDING](https://github.com/immutable-js/immutable-js/pull/1833/files#diff-07985fdcade0e64d11482724879a644f07879ba61b8fb6c6119e1b1902b72ae4) file. I do not know how does github funding works too, but actually, only @Methuselah96 is in here, @leebyron (at least) should probably be there too. 

Next steps are : 

- [ ] open [the remaining PR here](https://github.com/immutable-js-oss/immutable-js/pulls)
- [x] close the issues fixed by the fork
- [ ] delete the fork

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -606,40 +624,36 @@ Please contribute!
...
 and research in persistent data structures.
 
 [Hugh Jackson](https://github.com/hughfdjackson/), for providing the npm package
 name. If you're looking for his unsupported package, see [this repository](https://github.com/hughfdjackson/immutable).
 
+## License
 
-License
--------
+## License
```
Comment (bdurrer) [inline README.md:657]: double title, probably a mistake

File: __tests__/issues.ts [hunk 2]
```diff
@@ -94,3 +94,33 @@ describe('Issue #1293', () => {
     expect(secondState).toEqual(firstState);
   });
 });
+
+describe('Issue #1643', () => {
```
Comment (bdurrer) [inline __tests__/issues.ts:98]: are the issue numbers on the main repo the same? might be hard to track otherwise

File: __tests__/issues.ts [hunk 3]
```diff
@@ -94,3 +94,33 @@ describe('Issue #1293', () => {
...
+    }
+
+    it(`Collection#hashCode() should handle objects that return ${label} for valueOf`, () => {
+      const set = Set().add(new MyClass());
+      set.hashCode();
+    });
+  });
+});
+
+describe('Issue #1785', () => {
```
Comment (bdurrer) [inline __tests__/issues.ts:122]: are the issue numbers on the main repo the same? might be hard to track otherwise

**Review by bdurrer — COMMENTED** [review]
just some potential, minor things

File: __tests__/issues.ts (hunk 2 above)
Comment (jdeniau) [reply __tests__/issues.ts:98]: The two issues your pointed are in fact really the issues in immutable-js repository !

File: __tests__/Map.ts [hunk 4]
```diff
@@ -258,6 +258,12 @@ describe('Map', () => {
     expect(r).toBe(m);
   });
 
+  it('ensures iter is unmodified', () => {
+    const m = Map({ a: 1, b: 1 });
+    const r = m.map((value, key, iter) => 2 * iter.get(key));
+    expect(r.toObject()).toEqual({ a: 2, b: 2 });
+  });
+
```
Comment (acusti) [inline __tests__/Map.ts:266]: i remember noticing this when this PR was originally made; the test case description here seems to have no relation to the assertion. maybe the thing to do is:

```suggestion
  it('provides unmodified original collection as 3rd iter argument', () => {
    const m = Map({ a: 1, b: 1 });
    const r = m.map((value, key, iter) => {
      expect(iter).toEqual(m);
      return 2 * iter.get(key);
    });
    expect(r.toObject()).toEqual({ a: 2, b: 2 });
  });

```

File: __tests__/OrderedSet.ts [hunk 5]
```diff
@@ -59,4 +59,80 @@ describe('OrderedSet', () => {
...
+    const aNotB = a.subtract(b);
+    const aNotC = a.subtract(c);
+    const aNotD = a.subtract(d);
+
+    expect(aNotB.size).toBe(21);
+    expect(aNotC.size).toBe(22);
+    expect(aNotD.size).toBe(23);
+  });
+
+  it('updating a value with ".map()" should keep the set ordered', () => {
```
Comment (acusti) [inline __tests__/OrderedSet.ts:113]:
```suggestion
  it('keeps the Set ordered when updating a value with .map()', () => {
```

File: __tests__/OrderedSet.ts [hunk 6]
```diff
@@ -59,4 +59,80 @@ describe('OrderedSet', () => {
       'CD',
     ]);
   });
+
+  it('ensure that `subtract` works correcly (see https://github.com/immutable-js-oss/immutable-js/issues/139 )', () => {
```
Comment (acusti) [inline __tests__/OrderedSet.ts:63]:
```suggestion
  it('handles subtract when Set contains >=32 elements (https://github.com/immutable-js-oss/immutable-js/issues/139 )', () => {
```

File: __tests__/OrderedSet.ts [hunk 7]
```diff
@@ -59,4 +59,80 @@ describe('OrderedSet', () => {
...
+    // Set b to 0-24
+    for (let i = 0; i < 25; i++) {
+      b = b.add(i);
+    }
+    // Set c to 0-23
+    c = b.butLast();
+    // Set d to 0-22
+    d = c.butLast();
+
+    // Internal list resizing happens on the final remove when subtracting c from a
```
Comment (acusti) [inline __tests__/OrderedSet.ts:103]: the final remove is `a.subtract(d)`. assuming the rest is correct, this should be:
```suggestion
    // Internal list resizing happens on the final remove when subtracting d from a
```

File: __tests__/OrderedSet.ts [hunk 8]
```diff
@@ -59,4 +59,80 @@ describe('OrderedSet', () => {
...
+    const someOfThem = fillArray(partialCapacity);
+    expect(someOfThem.length).toBe(17);
+
+    const existingItems = OrderedSet(someOfThem).intersect(allItems);
+
+    expect(allItems.subtract(existingItems).size).toBe(15);
+    expect(allItems.subtract(existingItems).size + someOfThem.length).toBe(32);
+  });
+
+  it('ensure that `subtract` works correctly (see https://github.com/immutable-js-oss/immutable-js/issues/96 )', () => {
```
Comment (acusti) [inline __tests__/OrderedSet.ts:85]:
```suggestion
  it('handles consecutive subtract invocations (https://github.com/immutable-js-oss/immutable-js/issues/96 )', () => {
```

File: __tests__/OrderedSet.ts (hunk 6 above)
Comment (acusti) [reply __tests__/OrderedSet.ts:63]: now that i’ve reviewed more of this PR, i’m thinking maybe this test case should be moved to `__tests__/issues.ts`

File: __tests__/OrderedSet.ts (hunk 8 above)
Comment (acusti) [reply __tests__/OrderedSet.ts:85]: now that i’ve reviewed more of this PR, i’m thinking maybe this test case should be moved to `__tests__/issues.ts`

File: resources/jestResolver.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+const path = require('path');
+const pkg = require('../package.json');
+
+module.exports = (request, options) => {
+  if (request === 'immutable') {
+    if (process.env.CI) {
+      // In CI environment, let's test the real builded file to be sure that the build does is not broken
```
Comment (acusti) [inline resources/jestResolver.js:7]:
```suggestion
      // In CI environment, test the real built file to be sure that the build is not broken
```

File: resources/jestResolver.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+      return path.resolve(options.rootDir, pkg.main);
+    }
+
+    // in development mode, we want sourcemaps and live reload, etc, so let's point to the src/ directory
```
Comment (acusti) [inline resources/jestResolver.js:11]:
```suggestion
    // In development mode, we want sourcemaps, live reload, etc., so point to the src/ directory
```

File: __tests__/OrderedSet.ts (hunk 8 above)
Comment (jdeniau) [reply __tests__/OrderedSet.ts:85]: Personally I'm not a big fan of a file named `issues`. For me, things related to `OrderedSet` should be in the `OrderedSet` file.
I will change for your better comment though.

**Pull request merged**